### PR TITLE
Drive net device with polly IO event manager

### PR DESCRIPTION
**Pull request by iulianbarbu**
## Reason for This PR

Refactor net device and drive it with polly #1526 .

## Description of Changes

In this PR the net device is ported on polly IO event manager.
The changes contained by this PR are as follows:
- added edge triggered event type to polly
- aggregated the information from `NetEpollHandler`, `RxVirtio`, `TxVirtio`, `Net` into one single struct `Net`.
- implemented `handle_read` for `Net` device, which is now also a polly `EventHandler`, and can
handle events related to Net device (queue events, tap events, rate limiter events).
- provided a single execution path which registers the `Net` device and also handles the `PATCH` rate limiters request from microvm runtime.
- separated the Net device into specific and self contained components, which resembles the `Block` refactoring, which will help us to draw a more generic view of devices.
- ported net unit tests on polly.

Note: the commits are reflecting incremental changes in the refactoring (which might present changes to changes), but the overall changes are reflected in the commit description, so keep this in mind when reviewing.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided.
- [x] The description of changes is clear and encompassing.
- [ ] Docs have been provided.
- [ ] Code-level documentation for touched code is included in this PR.
- [x] No API changes are included in this PR.
- [x] The changes in this PR have no user impact.
- [x] No new `unsafe` code has been added.

### CODE REVIEW

**Comment by acatangiu** [comment]
@iulianbarbu I've edited this PR to be opened against the `refactoring/epoll_handler` branch, instead of the `master` branch.

File: src/vmm/src/builder.rs [hunk 1]
```diff
@@ -634,26 +628,28 @@ fn attach_net_devices(
...
-        ));
+            .map_err(StartMicrovmError::RegisterEvent)?;
 
         attach_mmio_device(
             vmm,
             cfg.iface_id.clone(),
-            MmioTransport::new(vmm.guest_memory().clone(), net_box)
-                .map_err(device_manager::mmio::Error::CreateMmioDevice)
-                .map_err(RegisterNetDevice)?,
+            MmioTransport::new(vmm.guest_memory().clone(), net_device.clone()).map_err(|e| {
```
Comment (acatangiu) [inline src/vmm/src/builder.rs]: No need to `clone` the `net_device` here.

File: src/devices/src/virtio/net.rs [hunk 2]
```diff
@@ -801,1154 +759,1039 @@ impl VirtioDevice for Net {
...
-        h.tx.queue_evt.write(1).unwrap();
-        h.handle_event(TX_QUEUE_EVENT, EPOLLIN).unwrap();
-
-        // Verify if TX queue was processed.
-        assert_eq!(txq.used.idx.get(), 1);
-        // Check if interrupt was triggered.
-        assert_eq!(h.interrupt_evt.read().unwrap(), 1);
-    }
-}
+//#[cfg(test)]
```
Comment (acatangiu) [inline src/devices/src/virtio/net.rs]: This PR needs the tests as well.

**Comment by acatangiu** [comment]
Overall it looks great! I will make another review pass and dive-deep to verify we're not losing any corner-case functionality.
:+1:

File: src/devices/src/virtio/net/device.rs [hunk 3]
```diff
@@ -520,6 +520,11 @@ impl Net {
 
     pub fn process_rx_queue_event(&mut self) {
         METRICS.net.rx_queue_event_count.inc();
+        if !self.queues[RX_INDEX].ready {
```
Comment (acatangiu) [inline src/devices/src/virtio/net/device.rs]: These checks are not enough. The `ready` flag on a queue does not involve that the queue is also fully configured and valid. See `queue.is_valid()` function.
Furthermore even if a particular queue is valid, the device might still not be fully activated.

Ideally we should check that the `device_activated` field from `struct MmioTransport` has been set.

Another option that doesn't need us to get a handle on the `MmioTransport` object, would be to verify that the `trait VirtioDevice::activate()` function implemented by the `NetDevice` object was called and returned success.

File: src/devices/src/virtio/block/device.rs [hunk 4]
```diff
@@ -94,6 +92,8 @@ pub struct Block {
     pub(crate) queue_evt: EventFd,
     mem: GuestMemory,
 
+    device_activated: bool,
```
Comment (raduiliescu) [inline src/devices/src/virtio/block/device.rs]: I would look at this from another perspective.
A device driver normally does its all internal initialization work and when is ready to receive data does the final "registration".
Here is case where the device is receiving data while is not ready and maybe this can be fixed by how the operations are sequenced.

One idea would be to create the TAP network device but not register it or mark it as IFF_DOWN and register it or make it IFF_UP when all is ready.

Comment (acatangiu) [reply src/devices/src/virtio/block/device.rs]: I hear you @raduiliescu, but I think that'd just make things more complicated than necessary. One of the simplifying assumptions that allowed us to simplify our architecture was that we register all events from the emulation thread.

> One idea would be to create the TAP network device but not register it or mark it as IFF_DOWN and register it or make it IFF_UP when all is ready.

- Registering the TAP fd with the event handler (listening for incoming packets) **after** the device is fully activated by the guest driver would mean doing said registration on a vcpu thread (vcpu threads do the virtio configuration since it's driven by the guest driver). We could do this by either:
  - wrap EventManager in `Arc<Mutex>` and keep references on both `vmm` and `vcpu` threads.
  - dispatch new "DeviceRegistrationComplete" events from the `vcpu` to the `vmm`, which will lead `vmm` to do the plugging of the device in EventManager.
- I believe marking the interface `IFF_DOWN/UP` might not be possible from within the jailed firecracker, even if it would, it's weird user experience to add a TAP to a microVM and have it flap up and down for a while out of your control.

Comment (sandreim) [reply src/devices/src/virtio/block/device.rs]: I agree with @acatangiu. We are definitely moving away from that pattern to simplify. Previously we were using a channel to inject the fds and handlers into the VMM thread.

I do not think the user would get a weird experience. It would make the experience more consistent if status of the tap device match the status of the guest nic.

Comment (raduiliescu) [reply src/devices/src/virtio/block/device.rs]: I was referring to the "register" that host TAP netdev is doing to the kernel when you create that interface, not to the register that Firecracker virtio net device is doing internally to the epoll handler. Anyway register was a sort of bad word :).
My idea here is to create the TAP, but don't let it receive packets. Allow it to receive packets when our driver is ready. One of the ways to do this is to set !IFF_UP when creating the TAP and set IFF_UP when Firecracker activates its virtio net device.

File: src/vmm/src/builder.rs (hunk 1 above)
Comment (acatangiu) [reply src/vmm/src/builder.rs]: nit: remove this `.clone()`

Comment (iulianbarbu) [reply src/vmm/src/builder.rs]: Done. Please take a look at the last force push.

File: src/devices/src/virtio/net/defs.rs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply src/devices/src/virtio/net/defs.rs]: Done. Please take a look at the last force push.

File: src/devices/src/virtio/net.rs (hunk 2 above)
Comment (iulianbarbu) [reply src/devices/src/virtio/net.rs]: Added unit tests in the last commit.

File: src/devices/src/virtio/net/device.rs (hunk 3 above)
Comment (iulianbarbu) [reply src/devices/src/virtio/net/device.rs]: Addressed this issue with an approach which enforces the device to implement `is_activated` and `set_device_activated` methods, bound to `VirtioDevice` trait.

File: src/devices/src/virtio/net/device.rs [hunk 6]
```diff
@@ -693,39 +697,268 @@ impl VirtioDevice for Net {
...
+        let mac = MacAddr::parse_str("11:22:33:44:55:66").unwrap();
+        let mut dummy = DummyNet::new(Some(&mac));
+        let n = dummy.net();
+
+        // Test `device_type()`.
+        {
+            assert_eq!(n.device_type(), TYPE_NET);
+        }
+
+        // // Test `queue_max_sizes()`.
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/net/device.rs]: nit: uncomment or delete this.

File: src/devices/src/virtio/net/device.rs [hunk 7]
```diff
@@ -693,39 +697,268 @@ impl VirtioDevice for Net {
...
+
+        {
+            let payload = frame_bytes_from_buf_mut(&mut frame_buf);
+            payload[0] = 15;
+        }
+        assert_eq!(frame_buf[vnet_hdr_len_], 15);
+    }
+
+    #[test]
+    #[allow(clippy::cognitive_complexity)]
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/net/device.rs]: Instead of allowing cognitive complexity, we should create smaller tests. From a quick look it seems like each block of code can be in its separate test as they don't depend on previous configurations.

File: src/devices/src/virtio/net/device.rs [hunk 8]
```diff
@@ -693,39 +697,268 @@ impl VirtioDevice for Net {
...
+            config_mac = [0u8; MAC_ADDR_LEN];
+            check_metric_after_block!(
+                &METRICS.net.cfg_fails,
+                1,
+                n.read_config(MAC_ADDR_LEN as u64 + 1, &mut config_mac)
+            );
+            assert_eq!(config_mac, [0u8, 0u8, 0u8, 0u8, 0u8, 0u8]);
+        }
+
+        // Let's test the activate function.
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/net/device.rs]: This should probably be deleted if we want to drop the activate function.

File: src/devices/src/virtio/net/device.rs [hunk 9]
```diff
@@ -16,7 +16,9 @@ use libc::EAGAIN;
 use logger::{Metric, METRICS};
 use memory_model::{GuestAddress, GuestMemory};
 use rate_limiter::{RateLimiter, TokenBucket, TokenType};
-use std::io::{Read, Write};
+#[cfg(not(test))]
+use std::io::Read;
+use std::io::Write;
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/net/device.rs]: nit: This should be imported in one go with `use std::io::{Read, Write};`

File: src/devices/src/virtio/net/device.rs [hunk 10]
```diff
@@ -86,6 +88,9 @@ pub struct Net {
     // Remove once MRG_RXBUF is supported and this variable is actually used.
     #[allow(dead_code)]
     mmds_ns: Option<MmdsNetworkStack>,
+
+    #[cfg(test)]
+    test_mutators: tests::TestMutators,
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/net/device.rs]: Since we create test mutators, why not also use them for checking bad_activate, reads & writes that we would otherwise test using the global metrics object? This way we can move a bit closer to the goal of running the tests in parallel.

File: src/devices/src/virtio/net/device.rs [hunk 11]
```diff
@@ -761,49 +834,103 @@ mod tests {
...
+
+            (rxq, txq)
+        }
+
+        fn set_mac(&mut self, mac: MacAddr) {
+            self.guest_mac = Some(mac);
+            let mut config_space;
+            config_space = Vec::with_capacity(MAC_ADDR_LEN);
+            // This is safe, because we know the capacity is large enough.
+            unsafe { config_space.set_len(MAC_ADDR_LEN) }
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/net/device.rs]: You can use the `vec!` macro instead which guarantees that it allocates exactly `MAC_ADDR_LEN` bytes. In this case capacity will be equal to len and you won't need to use the unsafe `set_len` function.

File: src/devices/src/virtio/net/event_handler.rs [hunk 12]
```diff
@@ -26,7 +27,10 @@ impl EventHandler for Net {
             _ if source == virtq_tx_ev_fd => self.process_tx_queue_event(),
             _ if source == rx_rate_limiter_fd => self.process_rx_rate_limiter_event(),
             _ if source == tx_rate_limiter_fd => self.process_tx_rate_limiter_event(),
-            _ => error!("Unknown event. No handling was done."),
+            _ => {
+                error!("Unknown event. No handling was done.");
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/net/event_handler.rs]: This message should be "unknown event source" or similar. The event is known, the source is the one that it's not.

File: src/devices/src/virtio/net/device.rs (hunk 9 above)
Comment (andreeaflorescu) [reply src/devices/src/virtio/net/device.rs]: Oh, I missed that. You're right, we can keep it as is now.

File: src/devices/src/virtio/net/device.rs (hunk 11 above)
Comment (andreeaflorescu) [reply src/devices/src/virtio/net/device.rs]: I think these lines also show somewhere else in the code. Can you also replace those pls?

File: src/devices/src/virtio/net/device.rs [hunk 13]
```diff
@@ -139,26 +54,127 @@ fn init_vnet_hdr(buf: &mut [u8]) {
...
     interrupt_evt: EventFd,
+
+    config_space: Vec<u8>,
+    guest_mac: Option<MacAddr>,
+
+    device_activated: bool,
+
     // TODO(smbarber): http://crbug.com/753630
     // Remove once MRG_RXBUF is supported and this variable is actually used.
     #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply src/devices/src/virtio/net/device.rs]: The `TODO(smbarber)` is also outdated, remove it.

File: src/devices/src/virtio/net/device.rs [hunk 14]
```diff
@@ -139,26 +54,123 @@ fn init_vnet_hdr(buf: &mut [u8]) {
...
+            rx_deferred_irqs: false,
+            rx_bytes_read: 0,
+            rx_frame_buf: [0u8; MAX_BUFFER_SIZE],
+            tx_frame_buf: [0u8; MAX_BUFFER_SIZE],
+            tx_iovec: Vec::with_capacity(QUEUE_SIZE as usize),
+            interrupt_status: Arc::new(AtomicUsize::new(0)),
+            interrupt_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            guest_mac: Some(MacAddr::from_bytes_unchecked(
+                &config_space.clone()[..MAC_ADDR_LEN],
+            )),
```
Comment (acatangiu) [inline src/devices/src/virtio/net/device.rs]: I believe this will crash for net devices with no configured MAC address.

In such a case, `config_space` is an empty vec, and here you're cloning it and addressing out of bounds.

Comment (iulianbarbu) [reply src/devices/src/virtio/net/device.rs]: Yes, you are right. I have to check first if the config_space is populated.

File: src/devices/src/virtio/net/device.rs [hunk 15]
```diff
@@ -209,8 +225,8 @@ impl NetEpollHandler {
     // if a buffer was used, and false if the frame must be deferred until a buffer
```
Comment (acatangiu) [inline src/devices/src/virtio/net/device.rs]: nit: change `self.rx.frame_buf` to `self.rx_frame_buf`

File: src/devices/src/virtio/net/device.rs [hunk 16]
```diff
@@ -497,242 +514,144 @@ impl NetEpollHandler {
...
-                event: other,
-            }),
+        } else {
+            match self.process_rx() {
+                Ok(()) => (),
+                Err(e) => {
+                    error!("{:?}", e);
+                    METRICS.net.event_fails.inc();
+                }
+            }
```
Comment (acatangiu) [inline src/devices/src/virtio/net/device.rs]: There is this identical `match` (where just the object being matched is different) in a lot of places, I suggest we add a simple static function:
```rust
fn report_err(err: &DeviceError) {
    error!("{:?}", e);
    METRICS.net.event_fails.inc();
}
```

and all these `match`es can then become:

```rust
self.process_rx().unwrap_or_else(report_err);
```

File: src/devices/src/virtio/net/device.rs [hunk 17]
```diff
@@ -497,242 +514,144 @@ impl NetEpollHandler {
...
+        // If the limiter is not blocked, continue transmitting bytes.
+        {
+            match self.process_tx() {
+                Ok(()) => (),
+                Err(e) => {
+                    error!("{:?}", e);
+                    METRICS.net.event_fails.inc();
+                }
+            }
+        } else {
```
Comment (acatangiu) [inline src/devices/src/virtio/net/device.rs]: nit: empty `else`

File: src/devices/src/lib.rs [hunk 18]
```diff
@@ -37,6 +38,13 @@ pub trait EpollHandler: AsAny + Send {
     fn handle_event(&mut self, device_event: DeviceEventT, evset: epoll::Events) -> Result<()>;
 }
 
+// Function used for reporting error in terms of logging
+// but also in terms of METRICS net event fails.
+pub fn report_net_event_fail(err: Error) {
```
Comment (acatangiu) [inline src/devices/src/lib.rs]: ultra nit:
```suggestion
pub(crate) fn report_net_event_fail(err: Error) {
```

File: src/devices/src/virtio/net/event_handler.rs [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+            _ => {
+                error!("Unknown event source.");
+                METRICS.net.event_fails.inc();
+            }
+        }
+
+        vec![]
+    }
+
+    fn init(&self) -> Vec<PollableOp> {
```
Comment (acatangiu) [inline src/devices/src/virtio/net/event_handler.rs]: nit: testing this `init` function is fairly simple. See `src/firecracker/src/metrics.rs` for an example.

File: src/devices/src/virtio/mmio.rs [hunk 20]
```diff
@@ -179,22 +177,20 @@ impl MmioTransport {
...
-                // check will fail and take the device into an unusable state.
-                if !self.device_activated && self.are_queues_valid() {
+                let device_activated = self.locked_device().is_activated();
+                if !device_activated && self.are_queues_valid() {
                     self.locked_device()
                         .activate(self.mem.clone())
                         .expect("Failed to activate device");
-
-                    self.device_activated = true;
+                    self.locked_device().set_device_activated(true);
```
Comment (sandreim) [inline src/devices/src/virtio/mmio.rs]: Why not set the activate flag in the activate() callback ? There 3 fns which talk about device activation: activate(), set_device_activated and is_activated().

File: src/devices/src/virtio/net/device.rs [hunk 21]
```diff
@@ -139,26 +54,123 @@ fn init_vnet_hdr(buf: &mut [u8]) {
...
+        tap.set_offload(
+            net_gen::TUN_F_CSUM | net_gen::TUN_F_UFO | net_gen::TUN_F_TSO4 | net_gen::TUN_F_TSO6,
+        )
+        .map_err(Error::TapSetOffload)?;
+
+        let vnet_hdr_size = vnet_hdr_len() as i32;
+        tap.set_vnet_hdr_size(vnet_hdr_size)
+            .map_err(Error::TapSetVnetHdrSize)?;
+
+        let mut avail_features = 1 << VIRTIO_NET_F_GUEST_CSUM
```
Comment (sandreim) [inline src/devices/src/virtio/net/device.rs]: Nit: I would define this as a constant: DEFAULT_NET_AVAIL_FEATURES.

File: src/devices/src/virtio/net/device.rs [hunk 22]
```diff
@@ -497,242 +508,102 @@ impl NetEpollHandler {
...
-                        // There might be enough budget now to send the frame.
-                        self.process_tx()
-                    }
-                    Err(e) => {
-                        METRICS.net.event_fails.inc();
-                        error!("Failed to get tx rate-limiter event: {:?}", e);
-                        Err(DeviceError::RateLimited(e))
-                    }
-                }
+        if let Err(e) = self.queue_evts[RX_INDEX].read() {
```
Comment (sandreim) [inline src/devices/src/virtio/net/device.rs]: Nit: I am unsure how the compiler/llvm optimizes this, but it looks like this path favors the error cases in terms of branching. I would test for Ok and then for Err if that makes sense.

File: src/devices/src/virtio/net/device.rs [hunk 23]
```diff
@@ -746,10 +617,10 @@ impl VirtioDevice for Net {
         &mut self.queues
     }
 
-    fn get_queue_events(&self) -> std::io::Result<Vec<EventFd>> {
-        let mut queue_evts_copy = Vec::new();
-        for evt in self.queue_evts.iter() {
-            queue_evts_copy.push(evt.try_clone()?);
+    fn get_queue_events(&self) -> result::Result<Vec<EventFd>, std::io::Error> {
```
Comment (sandreim) [inline src/devices/src/virtio/net/device.rs]: nit: std::io::Result<Vec<EventFd>> was shorter.

File: src/devices/src/virtio/net/device.rs [hunk 24]
```diff
@@ -746,10 +617,10 @@ impl VirtioDevice for Net {
...
+        let mut queue_evts_copy: Vec<EventFd> = Vec::new();
```
Comment (sandreim) [inline src/devices/src/virtio/net/device.rs]: nit :
```suggestion
        let mut queue_evts_copy = Vec::new();
```

File: src/devices/src/virtio/mmio.rs (hunk 20 above)
Comment (acatangiu) [reply src/devices/src/virtio/mmio.rs]: I think the idea was to remove `activate()` once all devices are ported to polly, and only keep `set_activated()`.

I believe the reasoning behind that is to also allow remote deactivation not just activation.

Comment (sandreim) [reply src/devices/src/virtio/mmio.rs]: To me it looks a little bit confusing. I would consider adding the set_device_activated once we make the decision of removing the  activate function. On top of that the semantics of reset() should set the device as inactive until reconfiguration completes.
